### ISSUE TITLE
SQL Join in numpy

### ISSUE THREAD

**Issue by zar1** [dssg/UPSG#25]
Pandas lets us do DataFrame.join. We probably need to provide some equivalent.

**Comment by zar1** [comment]
After conferring with @potash, it seems the right thing to do is probably to wrap the Numpy structured arrays in Pandas and use DataFrame.merge. Alternatives would be:

1) Write it myself. This would be tricky to implement correctly, and it would probably run slow
2) Ship data over to SQL and back. This requires a lot of unnecessary data copying, which would also make it slow. Although, it would at least certainly be correct.

**Comment by rayidghani** [comment]
May be worth providing but highly encouraging joins to happen in db.

**Comment by zar1** [comment]
Just to be sure we're on the same page @rayidghani, data in UPSG can effectively be stored in one of two ways:
1. In an SQL db
2. In a Numpy array saved in an HDF5 file

And we can transparently move between the two of them according to the needs of the user/application.

In the first case, it would definitely be preferable to join in the DB. The decision I'm trying to make is what happens in the second case. There doesn't appear to be a good way in pure Numpy to do a join-like operation on two Numpy arrays. Here, we get into the situation where we can either use Pandas, move the data to SQL and use that, or hand-roll something. Are you saying that you encourage us to use SQL if we're joining Numpy arrays, or just when things are located in SQL already?

**Comment by zar1** [comment]
I stumbled onto this:
http://pyopengl.sourceforge.net/pydoc/numpy.lib.recfunctions.html#-join_by

Of course, it exists in the mysterious lib.recfunctions.

**Comment by zar1** [comment]
Actually, recfunctions.join_by is pretty worthless. It has incorrect semantics if the arrays have a different number of rows. For example:

``` python
import numpy as np
from numpy.lib.recfunctions import join_by

a1 = np.array([(0, 'Lisa', 2),
                       (1, 'Bill', 1),
                       (2, 'Fred', 2),
                       (3, 'Samantha', 2),
                       (4, 'Augustine', 1),
                       (5, 'William', 0)], dtype=[('emp_id', int),
                                                  ('emp_name', 'S64'),
                                                  ('dept_id', int)])
a2 = np.array([(0, 'accts receivable'),
                       (1, 'accts payable'),
                       (2, 'shipping')], dtype=[('dept_id', int),
                                                ('dept_name', 'S64')])
joined = join_by('dept_id', a1, a2)
print joined
print joined.dtype
```

gives:

``` python
[(0, 5, 'William', 'accts receivable') (1, 1, 'Bill', 'accts payable')
 (2, 0, 'Lisa', --) (2, 2, 'Fred', --) (2, 3, 'Samantha', --)
 (2, 4, 'Augustine', 'shipping')]
[('dept_id', '<i8'), ('emp_id', '<i8'), ('emp_name', 'S64'), ('dept_name', 'S64')]
```

when we would expect:

``` python
[(0, 5, 'William', 'accts receivable') (1, 1, 'Bill', 'accts payable')
 (2, 0, 'Lisa', 'shipping') (2, 2, 'Fred', 'shipping') (2, 3, 'Samantha', 'shipping')
 (1, 4, 'Augustine', 'accts payable')]
[('dept_id', '<i8'), ('emp_id', '<i8'), ('emp_name', 'S64'), ('dept_name', 'S64')]
```

so recfunctions.join_by isn't really usable. Looks like pandas is the best option again

**Comment by zar1** [comment]
Pandas version implemented in 37363666c3
